x-pack/plugins/ingest_manager/server/plugin.ts

```diff
@@ -90,12 +92,31 @@ const allSavedObjectTypes = [
   ENROLLMENT_API_KEYS_SAVED_OBJECT_TYPE,
 ];
 
+/**
+ * Callbacks supported by the Ingest plugin
+ */
+export type ExternalCallbacks = [
+  'datasourceCreate',
+  (newDatasource: NewDatasource) => Promise<NewDatasource>
+];
+
+export type ExternalCallbacksStorage = Map<ExternalCallbacks[0], Set<ExternalCallbacks[1]>>;
+
 /**
  * Describes public IngestManager plugin contract returned at the `startup` stage.
  */
 export interface IngestManagerStartContract {
   esIndexPatternService: ESIndexPatternService;
   agentService: AgentService;
+  /**
+   * Services for Ingest's Datasources
+   */
+  datasourceService: typeof datasourceService;
+  /**
+   * Register callbacks for inclusion in ingest API processing
+   * @param args
+   */
+  register: (...args: ExternalCallbacks) => void;
```

Comment:
`register` is too generic, especially as a top-level API, IMO.

How about `register[External]Callbacks`?

Reply:
Also why is it a top level service of the contract?

Reply:
I had created the method in a way that is not specific to any one single registration type, thus why the generic name was used. The input argument (first one) would determine what type of callback is being registered. 

That being said, i'm not against changing it. You all ok if we go with `registerExternalCallback()`?

Reply:
Yes, so we know it has only one responsibility.

Reply:
I will rename it. Thanks for the feedback

re: `unregister()` 
I had that [question in the issue](https://github.com/elastic/kibana/issues/68698#issuecomment-644965487), but we could not identify a need for it, so I left it out for now